Bug fixes, optional cloudflare, almost free lnaddress (1 sat)

skrafft opened the pull request.

callebtc commented:
Hi! Thanks for this contribution. Please provide a thorough description of your PR.

callebtc reviewed (changes requested):
Please provide a thorough description of all changes introduces in this PR.

skrafft commented:
@callebtc Sorry, first PR here. The PR contains the followings:

- Bug fix: cannot send to lnaddresses with subdomains a@b.c.d but only to a@c.d (same fix as PR #914 that is not yet merged)
- Improvement: Do not require Cloudflare to manage an lnaddress domain and make clouflare token and zone_id optional. Updated documentation to describe how to setup DNS without cloudflare
- Bug fix: cannot make lnaddress free because of this (#549). So the minimum amount of an invoice is 1 sat.
- Bug fix: fixed refactoring done in this merge request because it broke lnaddresses (Bug report #1048)

Feel free to ask if you need more information.

callebtc commented:
Thanks that's awesome! 

However, we need to fix those things in separate PRs so we can be sure to know the cause if something breaks. It's better to work atomically, issue by issue. Would you be up for that and submit new PRs with the individual fixes separated from each other?

callebtc commented on lnbits/extensions/lnaddress/views_api.py:
```diff
@@ -67,15 +67,16 @@ async def api_domain_create(
         root_url = urlparse(str(request.url)).netloc
 
-        cf_response = await cloudflare_create_record(domain=domain, ip=root_url)
+        if domain.cf_token != "" and domain.cf_zone_id != "":
```
```suggestion
        if domain.cf_token and domain.cf_zone_id:
```

callebtc commented on lnbits/extensions/lnaddress/views_api.py:
```diff
@@ -150,7 +151,7 @@ async def api_lnaddress_make_address(
 
     domain_cost = domain.cost
-    sats = data.sats
+    sats = max(data.sats, 1) #Min amount is 1sat because 0sats invoice are not payable
```
Again, third time this appears.

My feeling is that the minimum 1 sat rule should be enforced in the frontend and there should be a check here that throws an error if it is not. It's not good practice to overwrite variables (adds confusion, hard to debug).

skrafft replied:
That's what I initially did (enforcing 1 sat on the frontend side) but this line: https://github.com/lnbits/lnbits/blob/5dda7eb2b9ad57b850a042aaeaf8d6165ac3ee2c/lnbits/extensions/lnaddress/views_api.py#L157 prevents it. I'm not really sure why it's there and did not want to risk refactoring it.

callebtc replied:
I would guess that this is a sanity check. Here, you could add the `data.sats == 0 or` check for example.

skrafft commented:
Sure, sorry. I'll do it now. I'll not open a PR for the same change as #914 as it's already there.

callebtc commented:
@skrafft sorry I got this extension now. We can reopen it but there are a few things that I would change.

skrafft commented on lnbits/extensions/lnaddress/crud.py:
```diff
@@ -83,7 +83,7 @@ async def create_address(
             data.wallet_key,
             data.wallet_endpoint,
-            data.sats,
+            max(data.sats,1),
```
1. Ok, I'll move it somewhere else.
2. It's necessary for display as the amount of the paid invoice when creating the address is displayed in the address list.

skrafft commented on lnbits/extensions/lnaddress/lnurl.py:
```diff
@@ -70,9 +71,8 @@ async def lnurl_callback(address_id, amount: int = Query(...)):
-                        await address.lnurlpay_metadata(domain=domain.domain)
-                    ).encode("utf-8"),
+                    "unhashed_description": metadata,
+                    "description_hash": hashlib.sha256(metadata.encode("utf-8")).hexdigest(),
```
Agreed

skrafft replied:
I just checked, encoding as utf-8 breaks json serialization. we should not encode unhashed_description for this call.

callebtc replied:
Sounds about right!

skrafft commented on lnbits/extensions/lnaddress/models.py:
```diff
@@ -9,8 +9,8 @@
     domain: str = Query(...)
-    cf_token: str = Query(...)
-    cf_zone_id: str = Query(...)
+    cf_token: str = Query("")
+    cf_zone_id: str = Query("")
```
To make cf_token and cf_zone_id optional. I'm new to this framework, is there a better option that I don't know?

callebtc replied:
Got it! This is fine then!

callebtc commented on lnbits/extensions/lnaddress/models.py:
```diff
@@ -9,8 +9,8 @@
     domain: str = Query(...)
-    cf_token: str = Query(...)
-    cf_zone_id: str = Query(...)
+    cf_token: str = Query("")
+    cf_zone_id: str = Query("")
```
Got it. This might work: 

```suggestion
    cf_token: Optional[str] = Query("")
    cf_zone_id: Optional[str] = Query("")
```

callebtc commented:
Please make sure to run `make format` once before pushing again.

skrafft commented:
@callebtc I just pushed the necessary changes (including make format)

motorina0 commented:
@skrafft regarding:

> Bug fix: cannot make lnaddress free because of this (https://github.com/lnbits/lnbits/issues/549). So the minimum amount of an invoice is 1 sat.

1. there is a validation error when trying to create a "free" address
![image](https://user-images.githubusercontent.com/2951406/203036034-a88aac64-32e8-4042-a5f5-ef3fede64013.png)

2. for public LNbits servers, how can we prevent that any user create any address?
 - a regular user could create a `legend.lnbits.com` domain, map it to its own wallet and start selling/giving away `@legend.lnbits.com` addresses. Is there anything in place preventing this?

motorina0 commented on lnbits/extensions/lnaddress/templates/lnaddress/display.html:
```diff
@@ -240,8 +240,10 @@ <h6 class="q-my-none">
         if (!dialog.data.duration) return 0
-        let sats = dialog.data.duration * parseInt('{{ domain_cost }}')
-        dialog.data.sats = parseInt(sats)
+        let sats = Math.max(
```
- `dialog.data.duration` is string, resulting in `NaN`
Suggestion:

```js
const sats = Math.max(parseInt(dialog.data.duration) * parseInt('{{ domain_cost }}'), 1)
```

skrafft commented:
@motorina0 While thinking of a solution for this: "for public LNbits servers, how can we prevent that any user create any address?", I noticed that this plugin allows anyone to create any domain even if it already exists and is in use by another wallet. I therefore suggests 2 things:

1. Check if a domain is currently in use and validated when creating/updating domains
2. Use an env variable LNADDRESS_RESTRICTED_DOMAINS that would contain a list of restricted domains for public servers like this: LNADDRESS_RESTRICTED_DOMAINS="legend.lnbits.com,yyyy.anotherdomain.com,zzz@yetanotherdomain.com"

Do you think that's enough ?

skrafft commented:
@motorina0 I just pushed the code for my previous comment. Feel free to tell me if something else is needed

skrafft commented:
@motorina0, @callebtc I know you have a lot to do, if you need more changes to merge this PR, just ask and I'll do it.

motorina0 commented:
> @motorina0, @callebtc I know you have a lot to do, if you need more changes to merge this PR, just ask and I'll do it.

Thanks, no changes required so far, but I plan to add some commits on top of this PR.

arcbtc commented:
The issues around having a solution based on list of public domains in .env far outweighs any drawbacks of just sticking with Cloudflare

skrafft commented:
@arcbtc Sorry, I'm not english native, I'm not sure I understand the meaning of this sentence. Is the solution bad ?

arcbtc commented:
Bad. Its better to just use Cloudflare

skrafft commented:
@arcbtc Except not everyone actually wants to use Cloudflare. Do you think it would be better if we added another method to check that you own the domain based on DNS TXT record. We could use dnspython to check a _lnbits.exampledomain.com TXT record which would contain: sha256(wallet_id + domain_name).

motorina0 commented:
> @arcbtc Sorry, I'm not english native, I'm not sure I understand the meaning of this sentence. Is the .env solution good or bad ?

@skrafft 
 - we are trying to keep to a minimum the number of configurations that a user has to make in order for spin-up an **LNbits** instance safely
 - even if the suggested solution looks simple in the context of this PR, it would slowly make the `.env` file very large if we take this path, therefore it is not optimal and will not be used
 - one option discussed here https://github.com/lnbits/lnbits/issues/1035#issuecomment-1321828639 is to allow `admin` users to manually add `addresses`. This works fine for small instances that don't use CloudFlare.

skrafft commented:
@motorina0 No problem, totally understandable. Manually adding addresses would definitely work for small instances. I have some spare time at the moment, tell me if there's anything I can do to help.

arcbtc commented:
Great idea @motorina0, and thanks for R&Ding this @skrafft. We have some "admin only" functionality in satspay extension now, perhaps this is a good excuse to do the same for lnaddress. Am admin could also potentially allow an automated lnaddress for their users. A bit like how WoS does, issue a random lnaddress to users.

arcbtc commented:
So admin toggles on "free lnaddress for all users" (in lnaddress as an admin feature), and when a user open lnaddress they get a free random lnaddress they can make use of

bitkarrot commented:
I think @motorina0 's suggestion is a good option for lightning addresses in LNbits on a limited use. I hope that the admin only option will be accessible through the API. 

If developers need more flexibliity, they can just simply use the API from lnurlp and build out their own LNaddress capability for multiple domains. (This is what we currently do on https://github.com/bitkarrot/laiseebot. Its ugly but it works) 

>     * one option discussed here [[Feature proposal] Lightning Address for LNURLp/wallet #1035 (comment)](https://github.com/lnbits/lnbits/issues/1035#issuecomment-1321828639) is to allow `admin` users to manually add `addresses`. This works fine for small instances that don't use CloudFlare.

Juju2323 reviewed (commented):
Let's get this fixed

callebtc commented:
Hi there, what's the status on this PR?

callebtc commented:
Please remove `package-lock.json` from the PR :)

skrafft commented:
Closing this PR since bugs have been fixed and addresses can now be created directly for free by admin. I'll open another PR for the lnaddress module to fix bugs I just found.
